src/java.base/share/classes/javax/crac/CheckpointException.java

```diff
@@ -36,6 +36,18 @@ public class CheckpointException extends Exception {
      * Creates a {@code CheckpointException}.
      */
     public CheckpointException() {
-        super();
+        // This exception is used as an aggregate for other (suppressed)
+        // exceptions; no point in recording and printing the stack trace.
+        super(null, null, true, false);
+    }
+
+    /**
+     * Constructs a {@code CheckpointException} with the specified
+     * detail message.
+     *
+     * @param message the detail message.
+     */
+    public CheckpointException(String message) {
```

Comment:
What if we remove this constructor and hide the other one? https://github.com/openjdk/crac/pull/60/files#r1190070472

Reply:
First of all, I think that `javax.crac` should mirror `jdk.crac` API- and docs-wise. It will be much easier when everyone will be able to just change the imports.

About the constructor with message: I find a bit confusing when an exception is thrown because of some problem with `criu` but there's no actionable message. I have added a simple 'Native checkpoint failed' but we should probably point user to the dump4.log file. (`criuengine` should also make some sanity checks on permissions but that's another thing). I wouldn't object to hiding it, though.

About the one without: `Context` narrows the `throws` to CE/RE and since we expect users to implement Context this would give them no chance to throw checked exceptions, not even the aggregating one. Hiding that won't work.

Reply:
Yes, that is a bug javax.crac does not follow jdk.crac.

Agree about the constructor without args.

And still want to delete the a constructor with message. We can introduce a jdk.crac.impl.CheckpointMessageException (along j.c.i.CheckpointOpenResourceException) which we use to communicate different reason(s) checkpoint is not successful.

Reply:
What about turning the inheritance the other way: final CheckpointAggregateException extends CheckpointException? This would be easily distinguishable, no need to change anything on Context interface. The no-arg constructor in CheckpointException would be protected.
It's kind of natural that exceptions carry messages.

Reply:
What will be the point of CheckpointException then? A more specific exception will also be preferable instead of that, isn't it?

Reply:
`Core` would throw a generic exception when something failed in the native part (`criuengine` returned exit code 1...). A custom implementation of `Context` would throw it when it can't call its children for some reason (but it's not a failure in the Resource itself).

Reply:
But that does not take away the need for the base CheckpointException to be able aggregate exceptions from Resources, so CheckpointException and Combined are mostly overlaping. Moreover,  so we'll need to be able to re-aggregate exceptions of two types. In the current form, Combined and CheckpointException are both fine to be thrown from a user context, and there is no guadance which one to use.

Let's keep it simple and use CheckpointException for aggregations as it was supposed to be. The message constructor should be removed, and every constructor-with-the-message call should be replaced with an exception of different type. Especially since we have examples for this [1].

[1] https://github.com/openjdk/crac/blob/6f403eaf7655f9ce4d10da25082a836d0ad1574c/src/java.base/share/classes/jdk/crac/Core.java#L164

Reply:
Thank you! I think the PR is mostly good, see the comments.